### PR TITLE
Add support for ChaCha20-Poly1305 with BoringSSL

### PR DESCRIPTION
**Pull request by facutuesca**
Since BoringSSL supports ChaCha20-Poly1305 through a different API than OpenSSL ([EVP_AEAD](https://commondatastorage.googleapis.com/chromium-boringssl-docs/aead.h.html) vs EVP), this change splits the AEAD backend into two: the original OpenSSL backend (`_aead_openssl.py`) and the new BoringSSL backend (`_aead_boringssl.py`).

The AEAD backend functions used by other modules (`aead._encrypt()`, `aead._decrypt()`, etc.) are now exposed through `aead/__init__.py` as wrappers. These wrappers select at runtime which backend to use. Currently only ChaCha20-Poly1305 + BoringSSL uses the BoringSSL backend.

This PR has two commits:
- The first one adds the bindings for BoringSSL's `EVP_AEAD` API
- The second one splits the old AEAD backend (`backends/openssl/aead.py`) into two, as described above, resulting in:
```
backends/openssl/aead
|-- __init__.py
|-- _aead_boringssl.py
`-- _aead_openssl.py # previously was aead.py
``` 

Fixes https://github.com/pyca/cryptography/issues/8903

### CODE REVIEW

**Comment by woodruffw** [comment]
At least one CI failure (this might need to be specialized for BoringSSL, if this test is newly being reached):

```
=================================== FAILURES ===================================
________________ TestOpenSSL.test_openssl_assert_error_on_stack ________________
[gw0] linux -- Python 3.9.16 /__w/cryptography/cryptography/.nox/tests/bin/python

self = <test_openssl.TestOpenSSL object at 0x7f89d4e70e20>

    def test_openssl_assert_error_on_stack(self):
        b = Binding()
        b.lib.ERR_put_error(
            b.lib.ERR_LIB_EVP,
            b.lib.EVP_F_EVP_ENCRYPTFINAL_EX,
            b.lib.EVP_R_DATA_NOT_MULTIPLE_OF_BLOCK_LENGTH,
            b"",
            -1,
        )
        with pytest.raises(InternalError) as exc_info:
            _openssl_assert(b.lib, False)
    
        error = exc_info.value.err_code[0]
>       assert error.lib == b.lib.ERR_LIB_EVP
E       AssertionError: assert 20 == 6
E        +  where 20 = <OpenSSLError(code=1682[96](https://github.com/pyca/cryptography/actions/runs/5026159816/jobs/9014094390?pr=8946#step:13:97)710, lib=20, reason=524550, reason_text=passed invalid argument)>.lib
E        +  and   6 = <module 'lib'>.ERR_LIB_EVP
E        +    where <module 'lib'> = <cryptography.hazmat.bindings.openssl.binding.Binding object at 0x7f89cecaffd0>.lib

tests/hazmat/bindings/test_openssl.py:78: AssertionError
```

**Comment by woodruffw** [comment]
Failure on `libressl` as well, possibly due to conflicting impls:

```
  warning: /home/runner/work/cryptography/cryptography/src/rust/target/release/build/cryptography-cffi-4a81e46800865770/out/_openssl.c:1010:14: error: conflicting types for ‘EVP_AEAD’; have ‘void’
  warning:  1010 | typedef void EVP_AEAD;
  warning:       |              ^~~~~~~~
  warning: In file included from /home/runner/work/cryptography/cryptography/osslcache/libressl-3.6.2-71b99442894e25741b4ac50b8185525317ad473a/include/openssl/cmac.h:62,
  warning:                  from /home/runner/work/cryptography/cryptography/src/rust/target/release/build/cryptography-cffi-4a81e46800865770/out/_openssl.c:641:
  warning: /home/runner/work/cryptography/cryptography/osslcache/libressl-3.6.2-71b99442894e25741b4ac50b8185525317ad473a/include/openssl/evp.h:1173:28: note: previous declaration of ‘EVP_AEAD’ with type ‘EVP_AEAD’ {aka ‘struct evp_aead_st’}
  warning:  1173 | typedef struct evp_aead_st EVP_AEAD;
  warning:       |                            ^~~~~~~~
  warning: /home/runner/work/cryptography/cryptography/src/rust/target/release/build/cryptography-cffi-4a81e46800865770/out/_openssl.c:1011:14: error: conflicting types for ‘EVP_AEAD_CTX’; have ‘void’
  warning:  1011 | typedef void EVP_AEAD_CTX;
  warning:       |              ^~~~~~~~~~~~
```

**Comment by alex** [comment]
It looks like LibreSSL actualy supports the same `EVP_AEAD_CTX` API as boring, so you may be able to just adjust the `ifdef` and have it work.

**Comment by alex** [comment]
The other CI failure should be resolved by rebasing on main. Once tests are green, let us know and we can review.

**Comment by facutuesca** [comment]
@woodruffw @alex Fixed the failures, all CI tests pass now

File: src/cryptography/hazmat/backends/openssl/aead/__init__.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    OpenSSL, so there is a separate code path for it
+    """
+    from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
+
+    return backend._lib.CRYPTOGRAPHY_IS_BORINGSSL and isinstance(
+        cipher, ChaCha20Poly1305
+    )
+
+
+def aead_cipher_supported(backend: Backend, cipher: _AEADTypes) -> bool:
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead/__init__.py:39]: Leading underscore, this shouldn't be public.

File: src/_cffi_src/openssl/evp_aead.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+int (*EVP_AEAD_CTX_seal)(const EVP_AEAD_CTX *, uint8_t *, size_t *, size_t,
+                         const uint8_t *, size_t, const uint8_t *, size_t,
+                         const uint8_t *, size_t) = NULL;
+int (*EVP_AEAD_CTX_open)(const EVP_AEAD_CTX *, uint8_t *, size_t *, size_t,
+                         const uint8_t *, size_t, const uint8_t *, size_t,
+                         const uint8_t *, size_t) = NULL;
+size_t (*EVP_AEAD_max_overhead)(const EVP_AEAD *) = NULL;
+EVP_AEAD_CTX *(*Cryptography_EVP_AEAD_CTX_new)(const EVP_AEAD *,
+                                               const uint8_t *, size_t,
+                                               size_t) = NULL;
```
Comment (alex) [inline src/_cffi_src/openssl/evp_aead.py:63]: When function pointers are nulled out, we add a constant and then add them to `_conditional.py` so they can't be used accidentally

File: src/cryptography/hazmat/backends/openssl/aead/__init__.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
+# This file is dual licensed under the terms of the Apache License, Version
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead/__init__.py:1]: No need for an `aead` subdirectory, these can just be `aead.py` and then the openssl/boringssl variants.

**Comment by alex** [comment]
It looks like this doesn't work on libressl -- did you investigate how complex it'd be to use this same API there?

**Comment by woodruffw** [comment]
Oh, I spoke too soon: `EVP_AEAD_CTX_new` creates a new context in LibreSSL, but defers initialization to `EVP_AEAD_CTX_init`. In BoringSSL, the two steps are combined.

**Comment by woodruffw** [comment]
Tests indicate that LibreSSL shows ChaCha20Poly1305 now, so I think I did that right. I'll do some more cleanup.

File: src/cryptography/hazmat/backends/openssl/aead/__init__.py (hunk 3 above)
Comment (woodruffw) [reply src/cryptography/hazmat/backends/openssl/aead/__init__.py:1]: Sounds good, will collapse into a single file.

Any thoughts on naming here? The "boring" impl now covers both BoringSSL and LibreSSL.

File: src/_cffi_src/openssl/evp_aead.py (hunk 2 above)
Comment (woodruffw) [reply src/_cffi_src/openssl/evp_aead.py:86]: Done here: https://github.com/pyca/cryptography/pull/8946/files#diff-660d86a2813941fb6224d37cade75ca2a7752ec37d662a7638625fd7e2df64beR264-R272

File: src/cryptography/hazmat/backends/openssl/aead/__init__.py (hunk 3 above)
Comment (alex) [reply src/cryptography/hazmat/backends/openssl/aead/__init__.py:1]: evp_aead?

Comment (woodruffw) [reply src/cryptography/hazmat/backends/openssl/aead/__init__.py]: Done; I've also prefixed the OpenSSL internals with `_openssl`, to keep things unambiguous.

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 4]
```diff
@@ -22,11 +22,216 @@
...
+        return _evp_aead_decrypt(
+            backend, cipher, nonce, data, associated_data, tag_length, ctx
+        )
+    else:
+        return _openssl_decrypt(
+            backend, cipher, nonce, data, associated_data, tag_length, ctx
+        )
+
+
+# EVP_AEAD support follows.
```
Comment (woodruffw) [inline src/cryptography/hazmat/backends/openssl/aead.py:111]: NB: I've added these comments for myself (to visually separate the internal impls), but can remove.

File: src/cryptography/hazmat/backends/openssl/aead.py [hunk 5]
```diff
@@ -256,57 +95,454 @@ def _decrypt(
...
-        data = data[:-tag_length]
-    if ctx is None:
-        cipher_name = _aead_cipher_name(cipher)
-        ctx = _aead_setup(
-            backend, cipher_name, cipher._key, nonce, tag, tag_length, _DECRYPT
+class _EVPAEAD:
+    """
+    Access to AEADs through EVP_AEAD APIs.
+    """
+
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/aead.py:112]: There's no need for a class here. Please just use a set of free functions.

File: src/cryptography/hazmat/bindings/openssl/_conditional.py [hunk 6]
```diff
@@ -125,7 +125,6 @@ def cryptography_has_tlsv13_functions() -> typing.List[str]:
         "SSL_set_post_handshake_auth",
         "SSL_SESSION_get_max_early_data",
         "SSL_write_early_data",
-        "SSL_read_early_data",
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/_conditional.py:128]: Why was this removed?

Comment (woodruffw) [reply src/cryptography/hazmat/bindings/openssl/_conditional.py:128]: Mistake! Will fix.

Comment (woodruffw) [reply src/cryptography/hazmat/bindings/openssl/_conditional.py]: [021f7ed](https://github.com/pyca/cryptography/pull/8946/commits/021f7ed13342e6a87f5edd0ceccf473f58503af8)

File: src/cryptography/hazmat/backends/openssl/aead.py (hunk 5 above)
Comment (woodruffw) [reply src/cryptography/hazmat/backends/openssl/aead.py]: [6ddc6bc](https://github.com/pyca/cryptography/pull/8946/commits/6ddc6bc8abf8e0958252ff3e4eccd87bbdcf4ea0) and [3849acd](https://github.com/pyca/cryptography/pull/8946/commits/3849acde8d7b759b2a6e5b25480f79f53337bc66)
